engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala

```diff
@@ -2793,6 +2793,7 @@ object IR {
       */
     sealed case class Self(
       override val location: Option[IdentifiedLocation],
+      synthetic: Boolean                          = false,
```

Comment:
As noted in the description we have explicit and implicit `self`, unfortunately. The primary reason for having this `boolean` is to distinguish between `self` explicitly defined as a parameter of a function and the `self` that is inserted by compiler in `GenerateMethodBodies` pass. We don't want the later to be available in the scope of the function so that whenever someone tries to do 
```
type Foo a b
    foo x = self.a + x
...
f = Foo 1 2
f.foo 3
```
it blows up, as it should.